subprojects/language-java/src/integTest/groovy/org/gradle/language/java/JavaLanguageIncrementalBuildIntegrationTest.groovy

```diff
@@ -38,4 +38,33 @@ class JavaLanguageIncrementalBuildIntegrationTest extends AbstractJvmLanguageInc
         executedAndNotSkipped ":compileMainJarMainJava", ":createMainJar", ":mainJar"
     }
 
-}
\ No newline at end of file
+    def "task outcome is up to date when no recompilation necessary"() {
+        given:
+        buildFile << '''
+            tasks.withType(org.gradle.api.tasks.compile.AbstractCompile) { task ->
+                task.options.incremental = true
+                task.inputs.file 'input.txt' 
+            }
+        '''.stripIndent()
+        file('input.txt') << 'first run'
+
+        when:
+        succeeds mainCompileTaskName
+
+        then:
+        executedAndNotSkipped mainCompileTaskName
+
+        when:
+        file('input.txt') << 'second run, triggers task execution, but no recompilation is necessary'
+
+        and:
+        succeeds mainCompileTaskName, '--debug'
+
+        then:
+        executed mainCompileTaskName
+        skipped mainCompileTaskName
+
+        and:
+        result.output.contains "Executing actions for task '${mainCompileTaskName}'."
```

Comment:
I think it would be enough that the compile task is up to date. If we implement this use case once via improved classpath snapshotting then we would not need to change the test. WDYT?